File: backend/services/cpf.py
"""
services/cpf.py — Qué es un CPF válido, y qué no.

POR QUE NO ALCANZA CON CONTAR ONCE DIGITOS

    El CPF lleva dos dígitos verificadores al final, calculados sobre los nueve
    primeros. O sea que de los cien mil millones de números de once cifras, sólo
    uno de cada cien es un CPF posible: la cuenta descarta el 99%.

    Hasta ahora la aplicación contaba once dígitos y nada más. Con eso,
    «12345678900» entra, y entra igual un teléfono tipeado en el campo
    equivocado. Y el CPF del pagador es lo que ata una recarga a la persona que
    la hizo: un número que no es de nadie no ata nada.

    El dedo que se equivoca en UN dígito es el caso que esto agarra siempre: al
    cambiar una cifra, la cuenta deja de dar. Es exactamente el error que más
    pasa y el que más caro sale, porque queda guardado y nadie lo mira hasta que
    hay que reclamar un pago.

LOS ONCE REPETIDOS SE RECHAZAN A PROPOSITO

    «11111111111», «00000000000» y los otros nueve pasan la cuenta de los
    verificadores —son un resultado legítimo de la fórmula— pero no existen como
    documento. Son los que aparecen cuando alguien quiere sacarse de encima un
    campo obligatorio, y el código de la aplicación ya usaba «00000000000» como
    valor por omisión para mandarle a Mercado Pago.

NO SE COMPRUEBA CONTRA LA RECEITA FEDERAL

    Esto dice que el número está bien formado, no que sea de quien lo escribe.
    Eso lo dice la foto del documento cuando se revisa el KYC, y es una decisión
    de una persona. Acá se corta lo que es imposible, que es barato y no falla
    nunca; lo otro necesita ojos.
"""
import re

# Los once dígitos repetidos. Se arma en vez de escribirse a mano porque la
# lista escrita a mano es la que termina con diez elementos y nadie lo nota.
_TODOS_IGUALES = {str(d) * 11 for d in range(10)}

_NO_DIGITO = re.compile(r"\D")


def normalizar(valor) -> str:
    """Los once dígitos, sin puntos ni guiones. Lo que se guarda y se compara.

    Se guarda normalizado y no como lo escribió la persona: «123.456.789-09» y
    «12345678909» son el mismo documento, y si se guardaran las dos formas la
    comparación de la recarga fallaría según cómo lo hubiera tipeado ese día.
    """
    return _NO_DIGITO.sub("", str(valor or ""))


def _digito(digitos: str, peso_inicial: int) -> str:
    """Un dígito verificador. La cuenta que define la Receita Federal."""
    suma = sum(int(d) * (peso_inicial - i) for i, d in enumerate(digitos))
    resto = (suma * 10) % 11
    # 10 y 11 valen cero. Es parte de la fórmula, no un caso borde nuestro.
    return "0" if resto >= 10 else str(resto)


def es_valido(valor) -> bool:
    """`True` si el número puede ser un CPF. No levanta: es para decidir."""
    n = normalizar(valor)
    if len(n) != 11 or n in _TODOS_IGUALES:
        return False
    return n[9] == _digito(n[:9], 10) and n[10] == _digito(n[:10], 11)


def para_mostrar(valor) -> str:
    """«123.456.789-09». Para la pantalla, nunca para guardar ni comparar."""
    n = normalizar(valor)
    if len(n) != 11:
        return n
    return f"{n[:3]}.{n[3:6]}.{n[6:9]}-{n[9:]}"


def tapado(valor) -> str:
    """«•••.456.789-••», para mostrarlo sin publicarlo entero.

    Se dejan los seis del medio: alcanzan para que la persona reconozca cuál de
    sus documentos es, y no alcanzan para que alguien que mira la pantalla por
    encima del hombro se lo lleve.
    """
    n = normalizar(valor)
    if len(n) != 11:
        return ""
    return f"•••.{n[3:6]}.{n[6:9]}-••"

File: frontend/src/utils/cpf.js
/**
 * cpf.js — Qué es un CPF válido, del lado del navegador.
 *
 * ES LA MISMA CUENTA QUE `backend/services/cpf.py`, Y ESTA EN LOS DOS LADOS
 *
 *   El servidor es el que decide: acá se puede apagar con la consola abierta,
 *   y por eso el registro y la recarga lo vuelven a comprobar allá. Esto está
 *   para que la persona se entere ANTES de mandar el formulario, no después.
 *
 *   Si las dos cuentas se separan, lo que pasa es que el navegador acepta algo
 *   que el servidor rechaza —o al revés—, y el cliente ve un error que no
 *   entiende. Son quince líneas idénticas a propósito.
 *
 * POR QUE NO ALCANZA CON CONTAR ONCE DIGITOS
 *
 *   El CPF lleva dos dígitos verificadores calculados sobre los nueve
 *   primeros: de cada cien números de once cifras, sólo uno es un CPF posible.
 *   Contar once acepta un teléfono tipeado en el campo equivocado, y acepta
 *   cualquier dedo que se equivoque en una cifra — que es el error que más
 *   pasa y el que queda guardado hasta que hay que reclamar un pago.
 */

/** Los once dígitos repetidos pasan la cuenta pero no son documentos de nadie. */
const TODOS_IGUALES = new Set(
  Array.from({ length: 10 }, (_, d) => String(d).repeat(11)),
);

/** Los dígitos, sin puntos ni guiones. Es lo que se manda al servidor. */
export function normalizarCpf(valor) {
  return String(valor || '').replace(/\D/g, '');
}

function digitoVerificador(digitos, pesoInicial) {
  let suma = 0;
  for (let i = 0; i < digitos.length; i += 1) {
    suma += Number(digitos[i]) * (pesoInicial - i);
  }
  const resto = (suma * 10) % 11;
  // 10 y 11 valen cero: es parte de la fórmula, no un caso borde nuestro.
  return resto >= 10 ? '0' : String(resto);
}

/** `true` si el número puede ser un CPF. */
export function cpfEsValido(valor) {
  const n = normalizarCpf(valor);
  if (n.length !== 11 || TODOS_IGUALES.has(n)) return false;
  return n[9] === digitoVerificador(n.slice(0, 9), 10)
      && n[10] === digitoVerificador(n.slice(0, 10), 11);
}

/** «123.456.789-09», para el campo. Nunca para mandar ni comparar. */
export function formatearCpf(valor) {
  const n = normalizarCpf(valor).slice(0, 11);
  if (n.length <= 3) return n;
  if (n.length <= 6) return `${n.slice(0, 3)}.${n.slice(3)}`;
  if (n.length <= 9) return `${n.slice(0, 3)}.${n.slice(3, 6)}.${n.slice(6)}`;
  return `${n.slice(0, 3)}.${n.slice(3, 6)}.${n.slice(6, 9)}-${n.slice(9)}`;
}

/**
 * El mensaje de lo que está mal, o cadena vacía si está bien.
 *
 * Se devuelve el TEXTO y no un booleano porque los dos motivos de rechazo se
 * arreglan distinto: «te falta un dígito» y «ese número no existe» mandan a la
 * persona a mirar cosas diferentes.
 */
export function queLeFaltaAlCpf(valor) {
  const n = normalizarCpf(valor);
  if (!n) return 'Ingresá tu CPF.';
  if (n.length !== 11) return 'El CPF tiene once dígitos.';
  if (!cpfEsValido(n)) return 'Ese CPF no es válido. Revisá los números.';
  return '';
}
